storage/id_alloc.go

```diff
@@ -54,21 +57,30 @@ func NewIDAllocator(idKey proto.Key, db *client.KV, minID int64, blockSize int64
 		minID:     minID,
 		blockSize: blockSize,
 		ids:       make(chan int64, blockSize+blockSize/2+1),
+		err:       make(chan error),
 	}
 	ia.ids <- allocationTrigger
-	return ia
+	return ia, nil
 }
 
-// Allocate allocates a new ID from the global KV DB. If multiple
-func (ia *IDAllocator) Allocate() int64 {
-	for {
-		id := <-ia.ids
-		if id == allocationTrigger {
-			go ia.allocateBlock(ia.blockSize)
-		} else {
-			return id
+// Allocate allocates a new ID from the global KV DB, and it will
+// start allocateBlock in background to prefetch ID
+func (ia *IDAllocator) Allocate() (id int64, err error) {
+	exit := false // flag to break loop while allocateBlock return error
+	// even error happens, return the allocated ID until channel is empty
+	for !exit || len(ia.ids) > 0 {
+		select {
+		case id = <-ia.ids:
+			if id == allocationTrigger {
+				go ia.allocateBlock(ia.blockSize)
+			} else {
+				return id, nil
+			}
+		case err = <-ia.err:
```

Comment:
If I'm reading this correctly, once an error has been returned, all future calls to Allocate will simply hang because there is nothing left to write to either the ids or err channels. If we cannot recover, then we need to save the error to return on all future Allocate calls. But an error handler that leaves the IDAllocator in a permanently broken state may be worse than a log.Fatal. I think we have to be able to retry on errors.